staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,750 @@
+type RequestDigest struct {
+	RequestInfo *request.RequestInfo
+	User        user.Info
+}
+
+// `*configController` implements Controller.  The methods of this
+// type and cfgMeal follow the convention that the suffix "Locked"
+// means that the caller must hold the configController lock.
+type configController struct {
+	queueSetFactory fq.QueueSetFactory
+
+	// configQueue holds TypedConfigObjectReference values, identifying
+	// config objects that need to be processed
+	configQueue workqueue.RateLimitingInterface
+
+	plLister         fclistersv1a1.PriorityLevelConfigurationLister
+	plInformerSynced cache.InformerSynced
+
+	fsLister         fclistersv1a1.FlowSchemaLister
+	fsInformerSynced cache.InformerSynced
+
+	flowcontrolClient fcclientv1a1.FlowcontrolV1alpha1Interface
+
+	// serverConcurrencyLimit is the limit on the server's total
+	// number of non-exempt requests being served at once.  This comes
+	// from server configuration.
+	serverConcurrencyLimit int
+
+	// requestWaitLimit comes from server configuration.
+	requestWaitLimit time.Duration
+
+	// This must be locked while accessing flowSchemas or
+	// priorityLevelStates.  It is the lock involved in
+	// LockingWriteMultiple.
+	lock sync.Mutex
+
+	// tracedValue holds the latest MatchingPrecedence from a
+	// FlowSchema with the tracer name.
+	tracedValue int32
+
+	// this condition is broadcast whenever tracedValue changes
+	traceCond sync.Cond
+
+	// flowSchemas holds the flow schema objects, sorted by increasing
+	// numerical (decreasing logical) matching precedence.  Every
+	// FlowSchema in this slice is immutable.
+	flowSchemas apihelpers.FlowSchemaSequence
+
+	// priorityLevelStates maps the PriorityLevelConfiguration object
+	// name to the state for that level.  Every name referenced from a
+	// member of `flowSchemas` has an entry here.
+	priorityLevelStates map[string]*priorityLevelState
+}
+
+// priorityLevelState holds the state specific to a priority level.
+type priorityLevelState struct {
+	// config holds the configuration after defaulting logic has been applied.
+	// Exempt may be true while there are queues, in the case of a priority
+	// level that recently switched from being non-exempt to exempt and whose
+	// queues are still draining.
+	// If there are queues then their parameters are here.
+	config fctypesv1a1.PriorityLevelConfigurationSpec
+
+	// qsCompleter holds the QueueSetCompleter derived from `config`
+	// and `queues` if config is not exempt, nil otherwise.
+	qsCompleter fq.QueueSetCompleter
+
+	// The QueueSet for this priority level.  This is nil if and only
+	// if the priority level is exempt.
+	queues fq.QueueSet
+
+	// quiescing==true indicates that this priority level should be
+	// removed when its queues have all drained.  May be true only if
+	// queues is non-nil.
+	quiescing bool
+
+	// number of goroutines between Controller::Match and calling the
+	// returned StartFunction
+	numPending int
```

Comment:
Because the computation of the flow distinguisher and the hash value is outside in the filter, between Match and the StartFunction.  I deliberately separated the two so that the lock would not be held while doing the SHA256.

Reply:
So a cursory internet search suggests that SHA256 consumes about two nanoseconds per byte hashed; I'm pretty sure this qualifies as a premature optimization.

Reply:
I added benchmarks.  Got this result:
```
mjs11:flowcontrol mspreitzus.ibm.com$ go test -bench .
goos: darwin
goarch: amd64
pkg: k8s.io/apiserver/pkg/util/flowcontrol
BenchmarkCRC-8   	19326795	        59.6 ns/op
BenchmarkSHA-8   	 3095313	       388 ns/op
PASS
ok  	k8s.io/apiserver/pkg/util/flowcontrol	3.985s
```